pkgs/build-support/bintools-wrapper/add-hardening.sh

```diff
@@ -1,33 +1,45 @@
-hardeningFlags=(relro bindnow)
-# Intentionally word-split in case 'hardeningEnable' is defined in
-# Nix. Also, our bootstrap tools version of bash is old enough that
-# undefined arrays trip `set -u`.
-if [[ -v hardeningEnable[@] ]]; then
-  hardeningFlags+=(${hardeningEnable[@]})
-fi
-hardeningLDFlags=()
+allHardeningFlags=(pie relro bindnow)
+hardeningFlags=()
 
-declare -A hardeningDisableMap
+declare -A hardeningEnableMap=()
 
-# Intentionally word-split in case 'hardeningDisable' is defined in Nix.
-for flag in ${hardeningDisable[@]:-IGNORED_KEY} @hardening_unsupported_flags@
-do
-  hardeningDisableMap[$flag]=1
+# Intentionally word-split in case 'NIX_HARDENING_ENABLE' is defined in Nix. The
+# array expansion also prevents undefined variables from causing trouble with
+# `set -u`.
+for flag in ${NIX_@infixSalt@_HARDENING_ENABLE-}; do
+  hardeningEnableMap[$flag]=1
 done
 
+# Remove unsupported flags.
 if (( "${NIX_DEBUG:-0}" >= 1 )); then
+  declare -A hardeningDisableMap=()
+fi
+for flag in @hardening_unsupported_flags@; do
+  [[ -n ${hardeningEnableMap[$flag]} ]] || continue
+  if (( "${NIX_DEBUG:-0}" >= 1 )); then
+    hardeningDisableMap[$flag]=1
```

Comment:
This only matters for `hardening_unsupported_flags` that are not in `allHardeningFlags`. Do we care?

Reply:
You mean "This only matters for hardening_unsupported_flags that *are* in allHardeningFlags", right? The idea is that if the toolchain doesn't support a given flag, we remove it from the set of enabled flags -- are you questioning the utility of that?

Reply:
Oh, you're drawing a comparison to the loop over `allHardeningFlags` below.

Reply:
Yes, the latter. Sorry that was unclear.

Reply:
I suppose we don't really need to report unsupported flags as if they were disabled. Should I remove this conditional (and its body)?